### ISSUE TITLE
[PROPOSAL] Open Work Process

### ISSUE THREAD

**Issue by artursouza** [dapr/community#45]
# Introduction

Dapr reached version 1.0 and as the community grows, there is a need for a more defined work process that is open to the community and still repeatable.

# Milestones

Each repo should track its milestones. Every issue being worked on should be tagged with the current milestone. If a pull request is not linked to an issue, then the PR should also be tagged with the current milestone.

# Planning

Planning should be done based on scope and not deadline. Maintainers should scan the backlog of each repo at least weekly and tag issues that should be worked on as part of the current milestone. Planning repos is done independently unless there is a new feature in runtime that requires coordination.

When contributors creates a pull request for an issue not planned for the current milestone, the maintainer of the repo must decide if the pull request will be kept on hold or if the issue will be added to the current milestone along with the PR.

## Projects

Major feature additions should be planned as a org-level project. Maintainers of all repos should agree on a sync up cadence and track work items and take note on a coordinated release date. The release date should not be treated like a hard deadline but as a tentative date for all the repos' maintainers to execute a coordinated release.

Sync up meetings for the project must be done via Zoom, opened to the community.

For instant messaging, it is recommended to have a dedicated channel in Dapr's Discord server for the project.

Projects should be short lived and focus on delivering a major feature and not for on-going planning.

# Versioning

Repositories must follow [Semantic Versioning v2.0](https://semver.org/spec/v2.0.0.html).

Each repository is versioned independently. For example, `dapr/dapr` might be at version `1.0` while `dapr/dotnet-sdk` might be at `1.1`.

# Release

Each repo (except `dapr/components-contrib`) can release independently. The exception, of course, is if there is a coordinated project in progress.

Each minor version upgrade must be tracked in a separate branch: `release-X.Y` where `X` is the major version and `Y` is the minor version.

## Hot-Fixes or Patches

A hot-fix must first be applied to the master branch and then cherry-picked into each `release-X.Y` that the fix is applicable.

**Comment by wcs1only** [comment]
Should we have a uniform versioning schema for RCs? We have at least three extant versioning schemas if I recall correctly.

**Comment by artursouza** [comment]
> Should we have a uniform versioning schema for RCs? We have at least three extant versioning schemas if I recall correctly.
> 
> Edit: I guess that is covered by the Semantic Versioning spec? We do not do this for all the SDKs today.

Yes, semantic versioning is the "schema". Which SDKs do not follow that?

**Comment by wcs1only** [comment]
> > Should we have a uniform versioning schema for RCs? We have at least three extant versioning schemas if I recall correctly.
> > Edit: I guess that is covered by the Semantic Versioning spec? We do not do this for all the SDKs today.
> 
> Yes, semantic versioning is the "schema". Which SDKs do not follow that?

python-sdk, dotnet-sdk, and components-contrib

**Comment by artursouza** [comment]
> > > Should we have a uniform versioning schema for RCs? We have at least three extant versioning schemas if I recall correctly.
> > > Edit: I guess that is covered by the Semantic Versioning spec? We do not do this for all the SDKs today.
> > 
> > 
> > Yes, semantic versioning is the "schema". Which SDKs do not follow that?
> 
> python-sdk, dotnet-sdk, and components-contrib

As long as we have Major.Minor.Patch right now, we can fix that going forward.

**Comment by msfussell** [comment]
Feedback
1) This channel needs to be created "For instant messaging, use the #release channel in Dapr's Discord server."
2) Reword "SDKs can release independently if needed, but should be avoided." - To "SDKs, CLI and dashboard repos can released independently if needed, but generally this should be done to address pressing issues that are specific to the repo, such as a high priority code fix".
3) Expand "All repositories release in a coordinated date." to "All repositories release with a regular cadence setting dates with a release tracking issue in dapr/dapr. Generally releases follow a 5-6 week schedule, allowing time at the start for the maintainers to plan what goes into a release based on existing issues and adapting throughout the milestone. The next release milestone planning issue with tentative dates is also published so that maintainers are able to indicate when to expect issues to be delivered that get moved out of the current milestone or longer lead items that need more than a single milestone to work on. The aim here is to relieve the expectation that every issue has to be included in the current milestone.  Each release has a tracking issue in dapr/dapr with the following template:" 
4) This whole section "A hot-fix must first be applied to the master branch and then cherry-picked into each release-X.Y that the fix is applicable. Hot-fixes should be released right away regardless of the coordinated release date". Needs clarity. See the hotfix description here https://docs.dapr.io/operations/support/support-release-policy/#introduction and that only 2 suppored releases are expected to hotfixed. Current and previous. Rather than repeating this here, maybe point to his document and make sure that we all agree to this. Believe me when the time is not far off when we are at v1.8 and someone asks to fix a bug in v1.1 and we will have to say, no, you need to upgrade to a supported version of Dapr. Each release adds a significant test/fix/release burden.
5) My only other thoughts are whether there is any need to align new labels or coordinate labels across repos. I think that generally we have this is good shape.

**Comment by msfussell** [comment]
Much of this proposal needs to be captured as the process for the release team, where the release team needs to be assigned and they are responsible for the overall release processes.

**Comment by artursouza** [comment]
Action: create a PR in the docs website.

**Comment by msfussell** [comment]
@artursouza - Can you close this issue https://github.com/dapr/community/issues/45 if the new release process document I wrote covers this? If not please update a PR on the release process now in this repo.
